File: Python/BOJ/3_Silver/1260.py
N, M, V = map(int, input().split())
edges = [set() for _ in range(N+1)]
for _ in range(M):
    a, b = map(int, input().split())
    edges[a].add(b)
    edges[b].add(a)

edges = list(map(sorted, edges))

result = []
visited = [False]*(N+1)


def dfs(node):
    visited[node] = True
    result.append(node)
    for next_node in edges[node]:
        if visited[next_node]:
            continue
        dfs(next_node)

    return result


def bfs(start_node):
    que = [start_node]
    visited = [False]*(N+1)
    visited[start_node] = True
    result = []

    while que:
        temp = []
        for node in que:
            result.append(node)

            for next_node in edges[node]:
                if visited[next_node]:
                    continue
                visited[next_node] = True
                temp.append(next_node)

        que = temp

    return result


print(*dfs(V))
print(*bfs(V))

File: JavaScript/BOJ/3_Silver/1260.js
const fs = require("fs");
const input = fs
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "../test.txt")
  .toString()
  .trim()
  .split("\n");

const [N, M, V] = input[0].split(" ").map((n) => parseInt(n));
const [, ...edgeInput] = input;

const edges = new Array(N + 1).fill(null).map(() => new Set());

for (let edge of edgeInput) {
  const [a, b] = edge
    .trim()
    .split(" ")
    .map((n) => parseInt(n));
  edges[a].add(b);
  edges[b].add(a);
}
edges.forEach((edge) => (edges[idx] = [...edge].sort((a, b) => a - b)));

const result = [];
const visited = new Array(N + 1).fill(false);

function dfs(currNode) {
  visited[currNode] = true;
  result.push(currNode);
  for (let nextNode of edges[currNode]) {
    if (visited[nextNode]) continue;
    dfs(nextNode);
  }

  return result;
}

function bfs(startNode) {
  let que = [startNode];
  const visited = new Array(N + 1).fill(false);
  visited[startNode] = true;
  const result = [];
  while (que.length > 0) {
    const tempQue = [];
    for (let currNode of que) {
      result.push(currNode);

      for (let nextNode of edges[currNode]) {
        if (visited[nextNode]) continue;
        visited[nextNode] = true;
        tempQue.push(nextNode);
      }
    }
    que = [...tempQue];
  }
  return result;
}

console.log(dfs(V).join(" "));
console.log(bfs(V).join(" "));
